include/perfetto/ext/base/sched.h

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright (C) 2025 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef INCLUDE_PERFETTO_EXT_BASE_SCHED_H_
+#define INCLUDE_PERFETTO_EXT_BASE_SCHED_H_
+
+#include "perfetto/base/build_config.h"
+
+#include <string>
+#include <tuple>
+
+#include "perfetto/base/logging.h"
+#include "perfetto/base/status.h"
+#include "perfetto/ext/base/no_destructor.h"
+#include "perfetto/ext/base/status_or.h"
+
+namespace perfetto::base {
+
+struct SchedConfig {
+  enum class SchedPolicy { kIdle, kBatch, kOther, kFifo, kRr };
+  static SchedConfig CreateIdle() { return {SchedPolicy::kIdle, 0, 0}; }
+  static SchedConfig CreateDefaultUserspacePolicy() { return CreateOther(0); }
+  static SchedConfig CreateOther(int nice) {
+    return {SchedPolicy::kOther, 0, nice};
+  }
+  static SchedConfig CreateBatch(int nice) {
+    return {SchedPolicy::kBatch, 0, nice};
+  }
+  static SchedConfig CreateFifo(unsigned int priority) {
+    return {SchedPolicy::kFifo, priority, 0};
+  }
+  static SchedConfig CreateRr(unsigned int priority) {
+    return {SchedPolicy::kRr, priority, 0};
+  }
+
+  SchedConfig(const SchedPolicy policy,
+              const unsigned int priority,
+              const int nice)
+      : policy_(policy), rt_priority_(priority), nice_(nice) {
+    switch (policy) {
+      case SchedPolicy::kIdle:
+        PERFETTO_DCHECK(priority == 0 && nice == 0);
+        break;
+      case SchedPolicy::kOther:
+      case SchedPolicy::kBatch:
+        PERFETTO_DCHECK(ValidateNiceValue(nice).ok() && priority == 0);
+        break;
+      case SchedPolicy::kFifo:
+      case SchedPolicy::kRr:
+        PERFETTO_DCHECK(ValidatePriority(priority).ok() && nice == 0);
+        break;
+    }
+  }
+
+  std::string ToString() const;
+
+  /**
+   * Return true if this priority is _lower_ than others'
+   * (that implies `this.KernelPriority()` is _higher_ than
+   * `other.KernelPriority()`).
+   *
+   * If both priorities have the same `KernelPriority()` value, we compare the
+   * policies.
+   *
+   * Strictly speaking, it is not correct to compare realtime priorities
+   * 'SCHED_FIFO' and 'SCHED_RR' based on their policy, but we do it for
+   * simplicity.
+   */
+  bool operator<(const SchedConfig& other) const {
+    const int inverted_priority =
+        -1 * static_cast<int>(InternalUnifiedPriority());
```

Comment:
Nothing drives me more mad than seeing functions being used only in one place.
What's the reason this InternalUnifiedPriority is a standalone function rather than being code inlined here? It hurts readability. Now if I want to understand waht this function does I have to scroll down and this interrupts my flow. Why?
Code gets read way more times than it gets written. please optimize your code for readability.

Reply:
> What's the reason this InternalUnifiedPriority is a standalone function rather than being code inlined here?

You suggest to replace it with a lambda, being placed inside the body of the `operator<` function? 
something like 
```
bool operator<(const SchedConfig& other) const {
    const auto inverted_prio = [](const SchedConfig& config) {
      if (config.policy_ == SchedPolicy::kIdle) {
        return -1 * (kKernelMaxPrio + 1);
      }
      return -1 * static_cast<int>(config.KernelPriority());
    };
    return std::forward_as_tuple(inverted_prio(*this), policy_) <
           std::forward_as_tuple(inverted_prio(other), other.policy_);
}
```
It seems to me not very idiomatic C++, TBH